Fix TypeError when CropBox is an indirect object reference

dhdaines opened the pull request:
There was a regression in the latest release as noted in #1004 - object references are everywhere! Beware!  Fixes #1004

There is a new test in `test_tools_pdf2txt.py` (hopefully it is getting run?) but also I tested it on Error.pdf from that issue.

**Checklist**

- [ x] I have read [CONTRIBUTING.md](https://github.com/pdfminer/pdfminer.six/blob/master/CONTRIBUTING.md). 
- [ x] I have added a concise human-readable description of the change to [CHANGELOG.md](https://github.com/pdfminer/pdfminer.six/blob/master/CHANGELOG.md).
- [ x] I have tested that this fix is effective or that this feature works.
- [ x] I have added docstrings to newly created methods and classes.
- [ x] I have updated the [README.md](https://github.com/pdfminer/pdfminer.six/blob/master/README.md) and the [readthedocs](https://github.com/pdfminer/pdfminer.six/tree/master/docs/source) documentation. Or verified that this is not necessary.

felixxm commented on pdfminer/pdfpage.py:
```diff
@@ -68,7 +68,8 @@ def __init__(
         mediabox_params: List[Any] = [
-            resolve1(mediabox_param) for mediabox_param in self.attrs["MediaBox"]
+            resolve1(mediabox_param)
+            for mediabox_param in resolve1(self.attrs["MediaBox"])
```
We sometimes hit `KeyError` in this line, e.g. `self.attrs`
```
{
    Parent: <PDFObjRef:2>,
    Type: /'Page'
}
```
Can we protect it against missing `MediaBox`?
```python
mediabox_params: List[Any] = []
if "MediaBox" in self.attrs:
    mediabox_params = [
        resolve1(mediabox_param)
        for mediabox_param in resolve1(self.attrs["MediaBox"])
    ]
```

dhdaines replied:
Indeed... MediaBox is required by the spec (at least for PDF 1.7) but as usual this doesn't mean that it will necessarily be present in random PDFs.

dhdaines replied:
Ah, I misspoke slightly - MediaBox is inheritable, so it isn't required on individual `Page`s, and `pdfminer` should always propagate it from the top-level `Pages` object.  If it's missing there, then we have bigger problems, do you have a test case where this is happening?

felixxm replied:
> do you have a test case where this is happening?

Unfortunately, I cannot share the PDF file :disappointed:

dhdaines replied:
No problem, in any case the code in this PR should be robust to `MediaBox` missing everywhere.  It is suprising that you would have a `Page` that fails to inherit it though, which might indicate some other bug in `pdfminer`.

chenxi-briink commented:
@felixxm it would be great if this PR could be merged soon so [pdfplumber](https://github.com/jsvine/pdfplumber/issues/1166) could depend on latest version of pdfminer. 

currently the pdfplubmer depends on `20231228`, which throws `TypeError: 'PDFObjRef' object is not iterable` for some other situations when extracting text from pages (sorry I couldn't provide more technical details since I'm not familiar with the internals)

felixxm commented:
> @felixxm it would be great if this PR could be merged soon so other libraries like [pdfplumber](https://github.com/jsvine/pdfplumber/issues/1166) could depend on latest version of pdfminer.

I'd love to see it merged, but I don't have such a power :superhero: I'm not a maintainer of this package.

raunakdoesdev commented:
+1 on this - would love to see it merged @pietermarsman

walterheck commented:
@pietermarsman would really like to see that release, do you have any plans for it? I see you have no github activity since August, which is a little concerning as maintainer of this repo :)

dhdaines commented:
In the meantime (shameless self-promotion) you could try [PAVÉS](https://github.com/dhdaines/paves) which implements a mostly pdfminer.six compatible interface and ought to be somewhat more robust (your mileage may vary).  It is also considerably faster on large documents due to the ability to parallelize layout analysis across multiple CPUs.

pietermarsman commented:
Similar changes where done in #1081 and I happened to process those first, so I merged this MR with master now. Will review it later.

pietermarsman commented:
@dhdaines I continued on your work to validate the MediaBox, CropBox and Contents. I refactored it into 3 parse methods. I hope that is ok with you. 

I also rewrote the changelog, because compared to master this is now fixing resolving the CropBox. 

If the changes look good to you, I'll merge it.

dhdaines commented:
> @dhdaines I continued on your work to validate the MediaBox, CropBox and Contents. I refactored it into 3 parse methods. I hope that is ok with you.
> 
> I also rewrote the changelog, because compared to master this is now fixing resolving the CropBox.

Yes, this looks good!

pietermarsman commented:
Thanks again @dhdaines for your time and contribution!
